File: estore-frontend/src/core/services/catalog.service.ts
import { API_ENDPOINTS, apiClient, toApiError, unwrapResponse } from './api';
import { Product, Category, PageResponse } from '@/shared/types';

interface BackendProductDTO {
  id: number;
  name: string;
  price: number;
  description: string;
  imageUrl?: string;
  imageUrls?: string;
  stockQuantity?: number;
  availableStock?: number;
  inStock?: boolean;
  lowStock?: boolean;
  categoryId?: number;
  categoryName?: string;
  active?: boolean;
  featured?: boolean;
}

const mapProduct = (product: BackendProductDTO): Product => ({
  id: product.id,
  name: product.name,
  price: Number(product.price),
  description: product.description,
  imageUrl: product.imageUrl || '',
  category: product.categoryId
    ? {
        id: product.categoryId,
        name: product.categoryName || 'Uncategorized',
        description: '',
        displayOrder: 0,
        active: true,
      }
    : undefined,
  categoryName: product.categoryName,
  categoryId: product.categoryId,
  inventory: {
    id: product.id,
    quantity: product.availableStock ?? product.stockQuantity ?? 0,
  },
  active: product.active ?? true,
  featured: product.featured,
  inStock: product.inStock,
  lowStock: product.lowStock,
  availableStock: product.availableStock ?? 0,
});

const mapProductPage = (page: PageResponse<BackendProductDTO>): PageResponse<Product> => ({
  ...page,
  content: page.content.map(mapProduct),
  number: page.page,
});

export const CatalogService = {
  getCategories: async (): Promise<Category[]> => {
    try {
      const response = await apiClient.get(API_ENDPOINTS.CATEGORIES);
      return unwrapResponse<Category[]>(response);
    } catch (error) {
      throw toApiError(error, 'Failed to fetch categories');
    }
  },

  getProducts: async (page = 0, size = 12): Promise<PageResponse<Product>> => {
    try {
      const response = await apiClient.get(API_ENDPOINTS.PRODUCTS, {
        params: { page, size },
      });
      const data = unwrapResponse<PageResponse<BackendProductDTO>>(response);
      return mapProductPage(data);
    } catch (error) {
      throw toApiError(error, 'Failed to fetch products');
    }
  },

  getProductById: async (id: number): Promise<Product> => {
    try {
      const response = await apiClient.get(API_ENDPOINTS.PRODUCT_DETAIL(id));
      const data = unwrapResponse<BackendProductDTO>(response);
      return mapProduct(data);
    } catch (error) {
      throw toApiError(error, 'Failed to fetch product');
    }
  },

  searchProducts: async (keyword: string, page = 0, size = 12): Promise<PageResponse<Product>> => {
    try {
      const response = await apiClient.get(API_ENDPOINTS.PRODUCT_SEARCH, {
        params: { q: keyword, page, size },
      });
      const data = unwrapResponse<PageResponse<BackendProductDTO>>(response);
      return mapProductPage(data);
    } catch (error) {
      throw toApiError(error, 'Failed to search products');
    }
  },

  getProductsByCategory: async (categoryId: number, page = 0, size = 12): Promise<PageResponse<Product>> => {
    try {
      const response = await apiClient.get(API_ENDPOINTS.PRODUCTS, {
        params: { categoryId, page, size },
      });
      const data = unwrapResponse<PageResponse<BackendProductDTO>>(response);
      return mapProductPage(data);
    } catch (error) {
      throw toApiError(error, 'Failed to fetch products');
    }
  },
};

File: estore-backend/src/main/java/com/estore/catalog/service/CatalogService.java
package com.estore.catalog.service;

import com.estore.catalog.dto.CategoryDTO;
import com.estore.catalog.dto.ProductDTO;
import com.estore.catalog.entity.Category;
import com.estore.catalog.entity.Product;
import com.estore.catalog.repository.CategoryRepository;
import com.estore.catalog.repository.ProductRepository;
import com.estore.exception.ResourceNotFoundException;
import com.estore.shared.dto.PageResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CatalogService {

    private final CategoryRepository categoryRepository;
    private final ProductRepository productRepository;
    private final com.estore.inventory.service.InventoryService inventoryService;

    @Transactional(readOnly = true)
    public List<CategoryDTO> getAllCategories() {
        return categoryRepository.findByActiveTrueOrderByDisplayOrderAsc()
                .stream()
                .map(this::toCategoryDTO)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public CategoryDTO getCategoryById(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category", "id", id));
        return toCategoryDTO(category);
    }

    @Transactional(readOnly = true)
    public PageResponse<ProductDTO> getProducts(int page, int size, Long categoryId) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<Product> productPage;

        if (categoryId != null) {
            productPage = productRepository.findByCategoryIdAndActiveTrue(categoryId, pageable);
        } else {
            productPage = productRepository.findByActiveTrue(pageable);
        }

        List<ProductDTO> products = productPage.getContent().stream()
                .map(this::toProductDTO)
                .collect(Collectors.toList());

        return PageResponse.<ProductDTO>builder()
                .content(products)
                .page(productPage.getNumber())
                .size(productPage.getSize())
                .totalElements(productPage.getTotalElements())
                .totalPages(productPage.getTotalPages())
                .first(productPage.isFirst())
                .last(productPage.isLast())
                .build();
    }

    @Transactional(readOnly = true)
    public ProductDTO getProductById(Long id) {
        Product product = productRepository.findByIdAndActiveTrue(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product", "id", id));
        return toProductDTO(product);
    }

    @Transactional(readOnly = true)
    public PageResponse<ProductDTO> searchProducts(String keyword, Long categoryId, int page, int size) {
        Pageable pageable = PageRequest.of(page, size);
        Page<Product> productPage;

        if (categoryId != null && keyword != null && !keyword.isBlank()) {
            productPage = productRepository.searchByKeywordAndCategory(keyword.trim(), categoryId, pageable);
        } else if (keyword != null && !keyword.isBlank()) {
            productPage = productRepository.searchByKeyword(keyword.trim(), pageable);
        } else if (keyword == null || keyword.isBlank()) {
            if (categoryId != null) {
                productPage = productRepository.findByCategoryIdAndActiveTrue(categoryId, pageable);
            } else {
                productPage = productRepository.findByActiveTrue(pageable);
            }
        } else {
            productPage = productRepository.findByActiveTrue(pageable);
        }

        List<ProductDTO> products = productPage.getContent().stream()
                .map(this::toProductDTO)
                .collect(Collectors.toList());

        return PageResponse.<ProductDTO>builder()
                .content(products)
                .page(productPage.getNumber())
                .size(productPage.getSize())
                .totalElements(productPage.getTotalElements())
                .totalPages(productPage.getTotalPages())
                .first(productPage.isFirst())
                .last(productPage.isLast())
                .build();
    }

    @Transactional(readOnly = true)
    public List<ProductDTO> getFeaturedProducts(int limit) {
        Pageable pageable = PageRequest.of(0, limit);
        return productRepository.findByFeaturedTrueAndActiveTrue(pageable)
                .stream()
                .map(this::toProductDTO)
                .collect(Collectors.toList());
    }

    private CategoryDTO toCategoryDTO(Category category) {
        return CategoryDTO.builder()
                .id(category.getId())
                .name(category.getName())
                .description(category.getDescription())
                .imageUrl(category.getImageUrl())
                .displayOrder(category.getDisplayOrder())
                .active(category.isActive())
                .build();
    }

    private ProductDTO toProductDTO(Product product) {
        com.estore.inventory.dto.InventoryDTO inventory;
        
        if (product.getInventory() != null) {
            // Use pre-fetched inventory to avoid N+1 queries
            var inv = product.getInventory();
            inventory = com.estore.inventory.dto.InventoryDTO.builder()
                    .availableQuantity(inv.getAvailableQuantity())
                    .inStock(inv.isInStock())
                    .lowStock(inv.isLowStock())
                    .build();
        } else {
            try {
                inventory = inventoryService.getInventoryByProductId(product.getId());
            } catch (Exception e) {
                // Fallback if inventory is missing
                inventory = com.estore.inventory.dto.InventoryDTO.builder()
                        .availableQuantity(product.getStockQuantity() != null ? product.getStockQuantity() : 0)
                        .inStock((product.getStockQuantity() != null ? product.getStockQuantity() : 0) > 0)
                        .lowStock(false)
                        .build();
            }
        }

        return ProductDTO.builder()
                .id(product.getId())
                .name(product.getName())
                .price(product.getPrice())
                .description(product.getDescription())
                .imageUrl(product.getImageUrl())
                .imageUrls(product.getImageUrls())
                .active(product.isActive())
                .featured(product.isFeatured())
                .stockQuantity(inventory.getAvailableQuantity())
                .availableStock(inventory.getAvailableQuantity())
                .inStock(inventory.isInStock())
                .lowStock(inventory.isLowStock())

                .categoryId(product.getCategory() != null ? product.getCategory().getId() : null)
                .categoryName(product.getCategory() != null ? product.getCategory().getName() : null)
                .createdAt(product.getCreatedAt())
                .build();
    }
}
